In this video we will explore how we can use various pandas techniques to handle the missing data in our datasets. We will learn how to find out how much data is missing and from which columns. We will see how to drop rows or columns where all or a lot of records are missing data. We will also learn how, instead of dropping data, we can also fill in the missing records with zeros or the mean of the remianing values

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
%cd /content/drive/My Drive/Colab Notebooks

/content/drive/My Drive/Colab Notebooks


In [3]:
import pandas as pd
data = pd.read_csv('data-titanic.csv')
data.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

It is a time to look at how many records are missing first. To do this we will first need to find out the total number of records in the dataset. We can do this by calling the shape property on the DataFrame.

In [4]:
data.shape

(891, 12)

We can see that the total number of records is 891 and that the total number of coluumns is 12.


We then find out the number of records in each column. We can do this by calling the count method on the DataFrame

In [5]:
data.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

The difference between the total records and the count per column represents the numnber of records missing from that column. Out of the 12 columns we have 3 columns where values are missing. For example. Age only has 714 values out of a total of 891 rows. Cabin only has 204 records and Embraked  has values for 889 records. There are different ways we could handle these missing values. One of the ways is to drop any row here a value is missing even from a single column.

In [6]:
data_missing_dropped = data.dropna()
data_missing_dropped.shape

(183, 12)

When we run this drop rows method, we assign the results back into a new DataFrame. This leaves us with just 183 records out of a total of 891. However, this may lead to losing a lot of the data and may not be acceptable.

Another method is to drop only those rows where all values are missing.

In [7]:
data_all_missing_dropped = data.dropna(how="all")
data_all_missing_dropped.shape

(891, 12)

We do this by setting the how parameter for the dropna method to all. 

Instead of dropping rows, another method is to fill in the missing values with some data. We can fill in the missing values with 0.

In [8]:
data_filled_zeros = data.fillna(0)
data_filled_zeros.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            891
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          891
Embarked       891
dtype: int64

Here we use pandas' fillna method, and we pass the numeric value of 0 to the column where the data should be filled in. You can see that we have now filled all teh missing values with 0 and hence the count for all the column has gone up to the total number of count of records in the dataset. 


Also , instead of filling in missing values with 0, we could fill them with the mean of the remaining exisiting values. To do this we call the fillna method on the column where we are filling the values in, and we pass the mean of the column as the parameter.

In [9]:
data_filled_in_mean = data.copy()
data_filled_in_mean.Age.fillna(data.Age.mean(), inplace=True)
data_filled_in_mean.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            891
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

For example here we did fill in the missing value of Agfe with the mean of the existing values.

**Extra work**

Try to apply all the above methods to fill in all the values to get 891 so we can elimiate all the missing values.